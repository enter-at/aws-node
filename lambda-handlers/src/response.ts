import {APIGatewayProxyResult} from 'aws-lambda';
import * as httpStatusCode from 'http-status-codes';
import {BadRequestError, ForbiddenError, InternalServerError, LambdaHandlerError, NotFoundError} from './error';

export function badRequest(description: string | undefined): APIGatewayProxyResult {
    const error: BadRequestError = new BadRequestError(description);
    return buildResult<BadRequestError>(error, httpStatusCode.BAD_REQUEST);
}

export function forbidden(description: string | undefined): APIGatewayProxyResult {
    const error: ForbiddenError = new ForbiddenError(description);
    return buildResult<ForbiddenError>(error, httpStatusCode.FORBIDDEN);
}

export function internalServerError(): APIGatewayProxyResult {
    const error: InternalServerError = new InternalServerError('InternalServerError');
    return buildResult<InternalServerError>(error, httpStatusCode.INTERNAL_SERVER_ERROR);
}

export function notFound(description: string | undefined): APIGatewayProxyResult {
    const error: NotFoundError = new NotFoundError(description);
    return buildResult<NotFoundError>(error, httpStatusCode.NOT_FOUND);
}

export function ok<T>(result: T): APIGatewayProxyResult {
    return buildResult<T>(result, httpStatusCode.OK);
}

export function created<T>(result: T): APIGatewayProxyResult {
    return buildResult<T>(result, httpStatusCode.CREATED);
}

export function noContent<T>(): APIGatewayProxyResult {
    return buildResult<null>(null, httpStatusCode.NO_CONTENT);
}

function buildResult<T>(result: T, statusCode: number): APIGatewayProxyResult {
    const body = result && JSON.stringify(result instanceof LambdaHandlerError
        ? {errors: result}
        : result
    );

    return {
        body,
        statusCode
    };
}
