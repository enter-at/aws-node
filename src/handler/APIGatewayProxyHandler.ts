import { APIGatewayProxyEvent, APIGatewayProxyResult } from "aws-lambda";
import {
    BadRequestError,
    ForbiddenError,
    FormatError,
    NotFoundError,
    RequestTimeoutError,
    ValidationError,
    UnauthorizedError,
} from "../error";
import { UnprocessableEntityError } from "../error/UnprocessableEntityError";
import { ContentTypeHeader, CORSHeader, Header, Headers } from "../header";
import logger from "../logger";
import {
    badRequest,
    forbidden,
    APIGatewayResponse,
    internalServerError,
    noContent,
    notFound,
    ok,
    requestTimeout,
    unauthorized,
    unprocessableEntity,
} from "../response";
import { BaseHandler, BaseHandlerArguments } from "./BaseHandler";

export interface APIGatewayProxyHandlerArguments extends BaseHandlerArguments {
    cors?: CORSHeader;
}

export class APIGatewayProxyHandler extends BaseHandler {
    private static handleError(err: Error): APIGatewayResponse {
        if (err instanceof ForbiddenError) {
            return forbidden(err.details);
        }
        if (err instanceof UnauthorizedError) {
            return unauthorized(err.details);
        }
        if (err instanceof BadRequestError || err instanceof FormatError || err instanceof ValidationError) {
            return badRequest(err.details);
        }
        if (err instanceof RequestTimeoutError) {
            return requestTimeout(err.details);
        }
        if (err instanceof UnprocessableEntityError) {
            return unprocessableEntity(err.details);
        }
        if (err instanceof NotFoundError) {
            return notFound(err.details);
        }
        logger.error({
            name: err.name,
            message: err.message,
            stack: err.stack,
        });
        return internalServerError();
    }

    private corsHeader: Header;

    constructor(args?: APIGatewayProxyHandlerArguments) {
        super(args);
        this.corsHeader = args?.cors ?? new CORSHeader("*", true);
    }

    protected after(result: APIGatewayResponse): APIGatewayProxyResult {
        result = result ?? noContent();
        if (result.statusCode === undefined) {
            result = ok(result);
        }
        return this.createResponse(result);
    }

    protected onException(exception: Error): APIGatewayProxyResult {
        return this.createResponse(APIGatewayProxyHandler.handleError(exception));
    }

    protected formatInput(event: APIGatewayProxyEvent): APIGatewayProxyEvent {
        if (!event.body) {
            return event;
        }
        try {
            event.body = this.inputFormat.apply(event.body);
            return event;
        } catch (err) {
            throw err;
        }
    }

    protected formatOutput(result: APIGatewayResponse): APIGatewayProxyResult {
        const { body, ...properties } = result;
        return {
            ...(body && { body: this.outputFormat.apply(body) }),
            ...properties,
        };
    }

    private createResponse(result: APIGatewayResponse): APIGatewayProxyResult {
        result.headers = this.createHeaders(result.headers);
        return this.formatOutput(result);
    }

    private createHeaders(headers: Headers | undefined): Headers {
        return {
            ...headers,
            ...(this.corsHeader && this.corsHeader.create()),
            ...(this.outputFormat && new ContentTypeHeader(this.outputFormat.contentType).create()),
        };
    }
}
