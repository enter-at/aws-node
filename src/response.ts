import { APIGatewayProxyResult } from "aws-lambda";
import { constants } from "http2";
import {
    BadRequestError,
    ForbiddenError,
    InternalServerError,
    LambdaHandlerError,
    NotFoundError,
    RequestTimeoutError,
    UnauthorizedError,
    UnprocessableEntityError,
} from "./error";
import { ConflictError } from "./error/ConflictError";

type Details = string | string[] | undefined;
export interface APIGatewayResponse extends Omit<APIGatewayProxyResult, "body"> {
    body: unknown | undefined;
}

export function badRequest(details: Details): APIGatewayResponse {
    const error: BadRequestError = new BadRequestError(details);
    return buildResult<BadRequestError>(error, constants.HTTP_STATUS_BAD_REQUEST);
}

export function forbidden(details: Details): APIGatewayResponse {
    const error: ForbiddenError = new ForbiddenError(details);
    return buildResult<ForbiddenError>(error, constants.HTTP_STATUS_FORBIDDEN);
}

export function unauthorized(details: Details): APIGatewayResponse {
    const error: UnauthorizedError = new UnauthorizedError(details);
    return buildResult<UnauthorizedError>(error, constants.HTTP_STATUS_UNAUTHORIZED);
}

export function internalServerError(): APIGatewayResponse {
    const error: InternalServerError = new InternalServerError("InternalServerError");
    return buildResult<InternalServerError>(error, constants.HTTP_STATUS_INTERNAL_SERVER_ERROR);
}

export function notFound(details: Details): APIGatewayResponse {
    const error: NotFoundError = new NotFoundError(details);
    return buildResult<NotFoundError>(error, constants.HTTP_STATUS_NOT_FOUND);
}

export function requestTimeout(details: Details): APIGatewayResponse {
    const error: RequestTimeoutError = new RequestTimeoutError(details);
    return buildResult<RequestTimeoutError>(error, constants.HTTP_STATUS_REQUEST_TIMEOUT);
}

export function unprocessableEntity(details: Details): APIGatewayResponse {
    const error: UnprocessableEntityError = new UnprocessableEntityError(details);
    return buildResult<UnprocessableEntityError>(error, constants.HTTP_STATUS_UNPROCESSABLE_ENTITY);
}

export function conflict(details: Details): APIGatewayResponse {
    const error: ConflictError = new ConflictError(details);
    return buildResult<ConflictError>(error, constants.HTTP_STATUS_CONFLICT);
}

export function ok<T>(result: T): APIGatewayResponse {
    return buildResult<T>(result, constants.HTTP_STATUS_OK);
}

export function created<T>(result: T): APIGatewayResponse {
    return buildResult<T>(result, constants.HTTP_STATUS_CREATED);
}

export function noContent(): APIGatewayResponse {
    return buildResult<null>(null, constants.HTTP_STATUS_NO_CONTENT);
}

function buildResult<T>(result: T, statusCode: number): APIGatewayResponse {
    const body =
        result && result instanceof LambdaHandlerError
            ? { errors: [{ name: result.name, details: result.details }] }
            : result;

    return {
        body,
        statusCode,
    };
}
