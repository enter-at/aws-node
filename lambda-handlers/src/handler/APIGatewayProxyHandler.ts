import {APIGatewayProxyEvent, APIGatewayProxyResult} from 'aws-lambda';
import {BadRequestError, FormatError, NotFoundError, ValidationError} from '../error';
import {ContentTypeHeader, CORSHeader, IHeader, IHeaders} from '../header';
import {badRequest, internalServerError, noContent, notFound, ok} from '../response';
import {BaseHandler, IBaseHandlerArguments} from './BaseHandler';

export interface IAPIGatewayProxyHandlerArguments extends IBaseHandlerArguments {
    cors?: CORSHeader;
}

export class APIGatewayProxyHandler extends BaseHandler {

    private static handleError(err: Error): APIGatewayProxyResult {
        if (err instanceof NotFoundError) {
            return notFound(err.description);
        }
        if (err instanceof BadRequestError || err instanceof FormatError || err instanceof ValidationError) {
            return badRequest(err.description);
        }
        return internalServerError();
    }

    private corsHeader: IHeader;

    constructor(args?: IAPIGatewayProxyHandlerArguments) {
        super(args);
        this.corsHeader = args?.cors ?? new CORSHeader('*', true);
    }

    protected after(result: any): APIGatewayProxyResult {
        if (result === undefined || result === null) {
            result = noContent();
        }
        if (result.statusCode === undefined) {
            result = ok(result);
        }
        return this.createResponse(result);
    }

    protected onException(exception: any): APIGatewayProxyResult {
        return this.createResponse(APIGatewayProxyHandler.handleError(exception));
    }

    protected formatInput(event: APIGatewayProxyEvent): APIGatewayProxyEvent {
        if (!event?.body) {
            return event;
        }
        try {
            event.body = super.formatInput(event.body);
            return event;
        } catch (err) {
            if (err instanceof FormatError) {
                throw new FormatError([{body: [err.description]}]);
            }
            throw err;
        }
    }

    protected formatOutput(result: APIGatewayProxyResult): APIGatewayProxyResult {
        if (result?.body) {
            result.body = super.formatOutput(result.body);
        } else {
            delete result.body;
        }
        return result;
    }

    private createResponse(result: APIGatewayProxyResult): APIGatewayProxyResult {
        result.headers = this.createHeaders(result.headers);
        return this.formatOutput(result);
    }

    private createHeaders(headers: IHeaders | undefined): IHeaders {
        return {
            ...headers,
            ...(this.corsHeader && this.corsHeader.create()),
            ...(this.outputFormat && new ContentTypeHeader(this.outputFormat.contentType).create())
        };
    }
}
