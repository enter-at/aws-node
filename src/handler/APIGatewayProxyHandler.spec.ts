import { APIGatewayProxyHandler } from "./APIGatewayProxyHandler";
import { APIGatewayProxyEvent, Handler, APIGatewayProxyResult, Context } from "aws-lambda";
import {
    RequestTimeoutError,
    InternalServerError,
    NotFoundError,
    ValidationError,
    FormatError,
    ForbiddenError,
    BadRequestError,
} from "../error";
import { ok, created, noContent } from '../response'
import * as ContextFactory from "../../test/fixtures/ContextFactory";
import * as APIGatewayProxyEventFactory from "../../test/fixtures/APIGatewayProxyEventFactory";

describe(APIGatewayProxyHandler.name, () => {
    let handler: APIGatewayProxyHandler;
    let context: Context;
    let event: APIGatewayProxyEvent;

    beforeEach(() => {
        handler = new APIGatewayProxyHandler();
        context = ContextFactory.factory();
        event = APIGatewayProxyEventFactory.factory();
    });

    it("handles HTTP 200 response correctly", async () => {
        const fn = handler.wrapper(() => {
            return ok({ result: "HTTP 200" })
        }) as Handler<APIGatewayProxyEvent, APIGatewayProxyResult>;

        const result = await fn(event, context, () => {});
        expect(result).toMatchSnapshot();
    });

    it("handles HTTP 201 response correctly", async () => {
        const fn = handler.wrapper(() => {
            return created({ result: "HTTP 201" })
        }) as Handler<APIGatewayProxyEvent, APIGatewayProxyResult>;

        const result = await fn(event, context, () => {});
        expect(result).toMatchSnapshot();
    });

    it("handles HTTP 204 response correctly", async () => {
        const fn = handler.wrapper(() => {
            return noContent()
        }) as Handler<APIGatewayProxyEvent, APIGatewayProxyResult>;

        const result = await fn(event, context, () => {});
        expect(result).toMatchSnapshot();
    });

    it("handles BadRequestError response correctly", async () => {
        const fn = handler.wrapper(() => {
            throw new BadRequestError("BadRequestError message");
        }) as Handler<APIGatewayProxyEvent, APIGatewayProxyResult>;

        const result = await fn(event, context, () => {});
        expect(result).toMatchSnapshot();
    });

    it("handles ForbiddenError response correctly", async () => {
        const fn = handler.wrapper(() => {
            throw new ForbiddenError("ForbiddenError message");
        }) as Handler<APIGatewayProxyEvent, APIGatewayProxyResult>;

        const result = await fn(event, context, () => {});
        expect(result).toMatchSnapshot();
    });

    it("handles FormatError response correctly", async () => {
        const fn = handler.wrapper(() => {
            throw new FormatError("FormatError message");
        }) as Handler<APIGatewayProxyEvent, APIGatewayProxyResult>;

        const result = await fn(event, context, () => {});
        expect(result).toMatchSnapshot();
    });

    it("handles InternalServerError response correctly", async () => {
        const fn = handler.wrapper(() => {
            throw new InternalServerError("InternalServerError message");
        }) as Handler<APIGatewayProxyEvent, APIGatewayProxyResult>;

        const result = await fn(event, context, () => {});
        expect(result).toMatchSnapshot();
    });

    it("handles NotFoundError response correctly", async () => {
        const fn = handler.wrapper(() => {
            throw new NotFoundError("NotFoundError message");
        }) as Handler<APIGatewayProxyEvent, APIGatewayProxyResult>;

        const result = await fn(event, context, () => {});
        expect(result).toMatchSnapshot();
    });

    it("handles RequestTimeoutError response correctly", async () => {
        const fn = handler.wrapper(() => {
            throw new RequestTimeoutError("RequestTimeoutError message");
        }) as Handler<APIGatewayProxyEvent, APIGatewayProxyResult>;

        const result = await fn(event, context, () => {});
        expect(result).toMatchSnapshot();
    });

    it("handles ValidationError response correctly", async () => {
        const fn = handler.wrapper(() => {
            throw new ValidationError("ValidationError message");
        }) as Handler<APIGatewayProxyEvent, APIGatewayProxyResult>;

        const result = await fn(event, context, () => {});
        expect(result).toMatchSnapshot();
    });
});
