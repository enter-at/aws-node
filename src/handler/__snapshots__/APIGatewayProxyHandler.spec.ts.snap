// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`APIGatewayProxyHandler handles BadRequestError response correctly 1`] = `
Object {
  "body": "{\\"errors\\":[{\\"name\\":\\"BadRequestError\\",\\"details\\":\\"BadRequestError message\\"}]}",
  "headers": Object {
    "Access-Control-Allow-Credentials": true,
    "Access-Control-Allow-Origin": "*",
    "Content-Type": "application/json",
  },
  "statusCode": 400,
}
`;

exports[`APIGatewayProxyHandler handles ForbiddenError response correctly 1`] = `
Object {
  "body": "{\\"errors\\":[{\\"name\\":\\"ForbiddenError\\",\\"details\\":\\"ForbiddenError message\\"}]}",
  "headers": Object {
    "Access-Control-Allow-Credentials": true,
    "Access-Control-Allow-Origin": "*",
    "Content-Type": "application/json",
  },
  "statusCode": 403,
}
`;

exports[`APIGatewayProxyHandler handles FormatError response correctly 1`] = `
Object {
  "body": "{\\"errors\\":[{\\"name\\":\\"BadRequestError\\",\\"details\\":\\"FormatError message\\"}]}",
  "headers": Object {
    "Access-Control-Allow-Credentials": true,
    "Access-Control-Allow-Origin": "*",
    "Content-Type": "application/json",
  },
  "statusCode": 400,
}
`;

exports[`APIGatewayProxyHandler handles HTTP 200 response correctly 1`] = `
Object {
  "body": "{\\"result\\":\\"HTTP 200\\"}",
  "headers": Object {
    "Access-Control-Allow-Credentials": true,
    "Access-Control-Allow-Origin": "*",
    "Content-Type": "application/json",
  },
  "statusCode": 200,
}
`;

exports[`APIGatewayProxyHandler handles HTTP 201 response correctly 1`] = `
Object {
  "body": "{\\"result\\":\\"HTTP 201\\"}",
  "headers": Object {
    "Access-Control-Allow-Credentials": true,
    "Access-Control-Allow-Origin": "*",
    "Content-Type": "application/json",
  },
  "statusCode": 201,
}
`;

exports[`APIGatewayProxyHandler handles HTTP 204 response correctly 1`] = `
Object {
  "headers": Object {
    "Access-Control-Allow-Credentials": true,
    "Access-Control-Allow-Origin": "*",
    "Content-Type": "application/json",
  },
  "statusCode": 204,
}
`;

exports[`APIGatewayProxyHandler handles InternalServerError response correctly 1`] = `
Object {
  "body": "{\\"errors\\":[{\\"name\\":\\"InternalServerError\\",\\"details\\":\\"InternalServerError\\"}]}",
  "headers": Object {
    "Access-Control-Allow-Credentials": true,
    "Access-Control-Allow-Origin": "*",
    "Content-Type": "application/json",
  },
  "statusCode": 500,
}
`;

exports[`APIGatewayProxyHandler handles NotFoundError response correctly 1`] = `
Object {
  "body": "{\\"errors\\":[{\\"name\\":\\"NotFoundError\\",\\"details\\":\\"NotFoundError message\\"}]}",
  "headers": Object {
    "Access-Control-Allow-Credentials": true,
    "Access-Control-Allow-Origin": "*",
    "Content-Type": "application/json",
  },
  "statusCode": 404,
}
`;

exports[`APIGatewayProxyHandler handles RequestTimeoutError response correctly 1`] = `
Object {
  "body": "{\\"errors\\":[{\\"name\\":\\"RequestTimeoutError\\",\\"details\\":\\"RequestTimeoutError message\\"}]}",
  "headers": Object {
    "Access-Control-Allow-Credentials": true,
    "Access-Control-Allow-Origin": "*",
    "Content-Type": "application/json",
  },
  "statusCode": 408,
}
`;

exports[`APIGatewayProxyHandler handles UnauthorizedError response correctly 1`] = `
Object {
  "body": "{\\"errors\\":[{\\"name\\":\\"UnauthorizedError\\",\\"details\\":\\"UnauthorizedError message\\"}]}",
  "headers": Object {
    "Access-Control-Allow-Credentials": true,
    "Access-Control-Allow-Origin": "*",
    "Content-Type": "application/json",
  },
  "statusCode": 401,
}
`;

exports[`APIGatewayProxyHandler handles UnprocessableEntityError response correctly 1`] = `
Object {
  "body": "{\\"errors\\":[{\\"name\\":\\"UnprocessableEntityError\\",\\"details\\":\\"UnprocessableEntityError message\\"}]}",
  "headers": Object {
    "Access-Control-Allow-Credentials": true,
    "Access-Control-Allow-Origin": "*",
    "Content-Type": "application/json",
  },
  "statusCode": 422,
}
`;

exports[`APIGatewayProxyHandler handles ValidationError response correctly 1`] = `
Object {
  "body": "{\\"errors\\":[{\\"name\\":\\"BadRequestError\\",\\"details\\":\\"ValidationError message\\"}]}",
  "headers": Object {
    "Access-Control-Allow-Credentials": true,
    "Access-Control-Allow-Origin": "*",
    "Content-Type": "application/json",
  },
  "statusCode": 400,
}
`;
